gluoncv/model_zoo/faster_rcnn/faster_rcnn.py

```diff
@@ -133,105 +156,90 @@ def hybrid_forward(self, F, x, gt_box=None):
-        # top_feat = F.Pooling(top_feat, global_pool=True, pool_type='avg', kernel=self._roi_size)
         top_feat = self.global_avg_pool(top_feat)
         cls_pred = self.class_predictor(top_feat)
-        box_pred = self.box_predictor(top_feat).reshape(
-            (-1, self.num_class, 4)).transpose((1, 0, 2))
+        box_pred = self.box_predictor(top_feat)
+        # cls_pred (B * N, C) -> (B, N, C)
+        cls_pred = cls_pred.reshape((self._max_batch, num_roi, self.num_class + 1))
+        # box_pred (B * N, C * 4) -> (B, N, C, 4)
+        box_pred = box_pred.reshape((self._max_batch, num_roi, self.num_class, 4))
 
         # no need to convert bounding boxes in training, just return
         if autograd.is_training():
-            box_pred = box_pred.transpose((1, 0, 2))
             return (cls_pred, box_pred, rpn_box, samples, matches,
                     raw_rpn_score, raw_rpn_box, anchors)
 
-        # translate bboxes
-        bboxes = self.box_decoder(box_pred, self.box_to_center(rpn_box)).split(
-            axis=0, num_outputs=self.num_class, squeeze_axis=True)
+        # cls_ids (B, N, C), scores (B, N, C)
         cls_ids, scores = self.cls_decoder(F.softmax(cls_pred, axis=-1))
+        # cls_ids, scores (B, N, C) -> (B, C, N) -> (B, C, N, 1)
+        cls_ids = cls_ids.transpose((0, 2, 1)).reshape((0, 0, 0, 1))
+        scores = scores.transpose((0, 2, 1)).reshape((0, 0, 0, 1))
+        # box_pred (B, N, C, 4) -> (B, C, N, 4)
+        box_pred = box_pred.transpose((0, 2, 1, 3))
+
+        # rpn_boxes (B, N, 4) -> B * (1, N, 4)
+        rpn_boxes = F.split(rpn_box, axis=0, num_outputs=self._max_batch, squeeze_axis=False)
+        # cls_ids, scores (B, C, N, 1) -> B * (C, N, 1)
+        cls_ids = F.split(cls_ids, axis=0, num_outputs=self._max_batch, squeeze_axis=True)
+        scores = F.split(scores, axis=0, num_outputs=self._max_batch, squeeze_axis=True)
+        # box_preds (B, C, N, 4) -> B * (C, N, 4)
+        box_preds = F.split(box_pred, axis=0, num_outputs=self._max_batch, squeeze_axis=True)
+
+        # per batch predict, nms, each class has topk outputs
         results = []
-        for i in range(self.num_class):
-            cls_id = cls_ids.slice_axis(axis=-1, begin=i, end=i+1)
-            score = scores.slice_axis(axis=-1, begin=i, end=i+1)
-            # per class results
-            per_result = F.concat(*[cls_id, score, bboxes[i]], dim=-1)
-
-            results.append(per_result)
-        result = F.concat(*results, dim=0).expand_dims(0)
-        if self.nms_thresh > 0 and self.nms_thresh < 1:
-            result = F.contrib.box_nms(
-                result, overlap_thresh=self.nms_thresh, topk=self.nms_topk,
-                id_index=0, score_index=1, coord_start=2)
-            if self.post_nms > 0:
-                result = result.slice_axis(axis=1, begin=0, end=self.post_nms).squeeze(axis=0)
+        for rpn_box, cls_id, score, box_pred in zip(rpn_boxes, cls_ids, scores, box_preds):
```

Comment:
concat class dim to batch dim, do NMS might be faster than loop.

Reply:
The loop is used for batch_size > 1. Each image goes through one nms with class dim as batch dim.